extensions/parquet/table/src/main/java/io/deephaven/parquet/table/layout/ParquetFlatPartitionedLayout.java

```diff
@@ -5,42 +5,61 @@
 
 import io.deephaven.engine.table.impl.locations.TableDataException;
 import io.deephaven.engine.table.impl.locations.impl.TableLocationKeyFinder;
+import io.deephaven.parquet.base.ParquetFileReader;
+import io.deephaven.parquet.base.ParquetFileReader.ParquetFileReaderException;
+import io.deephaven.parquet.base.util.LocalFSChannelProvider;
 import io.deephaven.parquet.table.location.ParquetTableLocationKey;
 import org.jetbrains.annotations.NotNull;
 
 import java.io.File;
 import java.io.IOException;
 import java.nio.file.DirectoryStream;
-import java.nio.file.Files;
+import java.nio.file.FileSystems;
 import java.nio.file.Path;
+import java.util.HashMap;
+import java.util.Map;
 import java.util.function.Consumer;
 
-import static io.deephaven.parquet.table.ParquetTableWriter.PARQUET_FILE_EXTENSION;
-
 /**
  * Parquet {@link TableLocationKeyFinder location finder} that will discover multiple files in a single directory.
  */
 public final class ParquetFlatPartitionedLayout implements TableLocationKeyFinder<ParquetTableLocationKey> {
 
+    private static ParquetTableLocationKey locationKey(Path path) {
+        return new ParquetTableLocationKey(path.toFile(), 0, null);
+    }
+
     private final File tableRootDirectory;
+    private final Map<Path, ParquetTableLocationKey> cache;
 
     /**
      * @param tableRootDirectory The directory to search for .parquet files.
      */
     public ParquetFlatPartitionedLayout(@NotNull final File tableRootDirectory) {
         this.tableRootDirectory = tableRootDirectory;
+        cache = new HashMap<>();
     }
 
     public String toString() {
         return ParquetFlatPartitionedLayout.class.getSimpleName() + '[' + tableRootDirectory + ']';
     }
 
     @Override
-    public void findKeys(@NotNull final Consumer<ParquetTableLocationKey> locationKeyObserver) {
-        try (final DirectoryStream<Path> parquetFileStream =
-                Files.newDirectoryStream(tableRootDirectory.toPath(), "*" + PARQUET_FILE_EXTENSION)) {
+    public synchronized void findKeys(@NotNull final Consumer<ParquetTableLocationKey> locationKeyObserver) {
+        try (final DirectoryStream<Path> parquetFileStream = FileSystems.getDefault().provider()
```

Comment:
Why do you prefer `FileSystems.getDefault().provider()` to `Files`? I think that's an incorrect choice, we should be using the directories FS, not the default.

Reply:
Given we have `File tableRootDirectory`, we _know_ that `FileSystems.getDefault()` will be used. That said, I'm happy to make this more generic in case this ever gets made more generic via `Path tableRootDirectory`.

Reply:
Looks like you went back to `Files`, with the appropriate overload (`java.nio.file.Files#newDirectoryStream(java.nio.file.Path, java.nio.file.DirectoryStream.Filter<? super java.nio.file.Path>)`), so I think I'm happy. I agree with your point that a `Path` made from `File.toPath` will always have the default FS.